tmt/steps/prepare/ansible.py

```diff
@@ -1,9 +1,29 @@
+import sys
+from typing import Any, List, Optional
+
 import click
 
 import tmt
-
-
-class PrepareAnsible(tmt.steps.prepare.PreparePlugin):
+import tmt.utils
+from tmt.steps.provision import Guest
+
+if sys.version_info >= (3, 8):
+    from typing import TypedDict
+else:
+    from typing_extensions import TypedDict
+
+# Type annotation for "data" package describing a guest instance. Passed
+# between load() and save() calls.
+StepDataType = TypedDict(
+    'StepDataType',
+    {
+        'playbook': str,
+        'playbooks': List[str]
```

Comment:
Actually `playbooks` is only kept for backward compatibility as we decided to keep all keys (almost) in singular form. So both variants can accept single playbook and multiple playbooks as [defined in the spec](https://tmt.readthedocs.io/en/latest/spec/plans.html#spec-plans-prepare-ansible). So I guess we should go with `Union` here?

Reply:
Interesting. Looking at the code, it would probably be better to change `playbook` to `List[str]`. `playbooks` is dropped from `data` at https://github.com/teemtee/tmt/blob/ef7b64523eb48dee43e72fa6a393d5bf4a6ad0cd/tmt/steps/prepare/ansible.py#L61

That's sort of normalization, but apparently, `playbooks` can be part of `data` at some point, even if it's effectively removed. I'd add a bit to do `if isinstance(data['playbook'], str): data['playbook'] = [data['playbook']]`, to normalize `playbook` to be a list of strings always, then `playbook` key would be declared as `List[str]`, without involving `Union` at all.

Reply:
> I'd add a bit to do `if isinstance(data['playbook'], str): data['playbook'] = [data['playbook']]`, to normalize `playbook` to be a list of strings always

This is done here using `listify()`:

https://github.com/teemtee/tmt/blob/43d250f267ff4c63af1530a46d92b1c55fe75859/tmt/steps/prepare/ansible.py#L68

Reply:
`playbook` is now `List[str]` and I added a conversion for when it's a single string. Please confirm that it's fine.

Reply:
As mentioned above, the conversion to list is already handled during `wake()`, no need to handle this twice.

Reply:
That looks like a wrong place for normalization: when this plugin gets instantiated for the first time, it can be given a data package with `playbook: "foo.yaml"`, therefore it must normalize `playbook` to match the announced type.

It is the first time `playbook` enters the plugin system, and I'd say the call to `listify` in `wake()` is actually not needed anymore, because `playbook` can never be `str` at that point. `playbook` turns into `List[str]` the first time data bundle enters the plugin. `save()`, `load()` and `wake()` should now all see `playbook` being a `List[str]` because nothing is going to change that value anymore (gets saved as a list => loaded as a list => `wake()` does not neet to apply `listify()`).

Reply:
Actually the [Plugin.wake()](https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L436) method was intended to take care of the data normalization because it can handle all three sources of data at one place:

* data from config files
* data from workdir (after wake up)
* command line options (both first time and after wake up)

When waking up a plugin it is possible to provide additional command line options which should override whatever was stored in the workdir. Thus the best place for data normalization is once these command line options are processed.

One tangible example is the `prepare --how shell` plugin:

https://github.com/teemtee/tmt/blob/a715aeefda29b4372e9a78cffffb906d35c94fb1/tmt/steps/prepare/shell.py#L52

Perhaps we could reconsider this. Maybe introduce a dedicated method `normalize()` to make this more obvious or delegate the task to `click` options handling? For the `shell` prepare example above it would be simple `mutliple=True` but I guess there might be much more complicated checks for some options.

Reply:
Yes, thanks.